Clang 9, lots of warnings about non-static variables in headers.

Long story short: Ozo has lots of template variables declared in header files where the value of the variable is the result of a type_traits's evaluation.

E.g.:

```
template <typename T>
constexpr auto OutputIterator = is_output_iterator<T>::value;
```

Clang is warning about this because a globally defined variable like this should either have an "extern" declaration before it, or be static. Otherwise from the compiler's perspective, you risk violating the one-definition-rule.

Replacing these template-variables with "using" declarations will resolve the complaint from clang.


Full text of warning follows.

> In file included from /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/connection_info.h:3:
> In file included from /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/connector.h:4:
> In file included from /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/connection.h:8:
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:28:16: warning: no previous extern declaration for non-static variable 'IsNone<std::chrono::duration<long, std::ratio<1, 1000000000> > >' [-Wmissing-variable-declarations]
> constexpr auto IsNone = is_none<T>::value;
>                ^
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:33:43: note: in instantiation of variable template specialization 'ozo::IsNone' requested here
> template <typename T, typename = Require<!IsNone<T>>>
>                                           ^
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:34:23: note: in instantiation of default argument for 'operator==<std::chrono::duration<long, std::ratio<1, 1000000000> > >' required here
> inline constexpr bool operator == (const T&, const none_t&) { return false;}
>                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/deadline.h:107:30: note: while substituting deduced template arguments into function template 'operator==' [with T = std::chrono::duration<long, std::ratio<1, 1000000000> >, $1 = (no value)]
>     return time_left(t, now) == time_traits::duration(0);
>                              ^
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:28:11: note: declare 'static' if the variable is not intended to be used outside of this translation unit
> constexpr auto IsNone = is_none<T>::value;
>           ^
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:28:16: warning: no previous extern declaration for non-static variable 'IsNone<ozo::detail::endian>' [-Wmissing-variable-declarations]
> constexpr auto IsNone = is_none<T>::value;
>                ^
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:33:43: note: in instantiation of variable template specialization 'ozo::IsNone' requested here
> template <typename T, typename = Require<!IsNone<T>>>
>                                           ^
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:34:23: note: in instantiation of default argument for 'operator==<ozo::detail::endian>' required here
> inline constexpr bool operator == (const T&, const none_t&) { return false;}
>                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/detail/endian.h:27:34: note: while substituting deduced template arguments into function template 'operator==' [with T = ozo::detail::endian, $1 = (no value)]
> constexpr Require<endian::native == endian::little, U> convert_to_big_endian(T value) noexcept {
>                                  ^
> /home/jonesmz/meshpp_frameworks/3rdparty/ozo/include/ozo/core/none.h:28:11: note: declare 'static' if the variable is not intended to be used outside of this translation unit
> constexpr auto IsNone = is_none<T>::value;
>           ^

--- a/include/ozo/connection.h
+++ b/include/ozo/connection.h
@@ -331,7 +331,7 @@ struct is_connection<connection<Ts...>> : std::true_type {};
 * @hideinitializer
 */
 template <typename T>
-constexpr auto Connection = is_connection<std::decay_t<decltype(unwrap_connection(std::declval<T>()))>>::value;
+inline constexpr auto Connection = is_connection<std::decay_t<decltype(unwrap_connection(std::declval<T>()))>>::value;
 
 /**
  * @defgroup group-connection-functions Related functions
@@ -681,7 +681,7 @@ struct connection_source_traits {
  * @ingroup group-connection-concepts
  */
 template <typename T>
-constexpr auto ConnectionSource = is_connection_source<std::decay_t<T>>::value;
+inline constexpr auto ConnectionSource = is_connection_source<std::decay_t<T>>::value;
 
 template <typename Provider, typename TimeConstraint, typename Handler>
 constexpr auto async_get_connection(Provider&& p, TimeConstraint t, Handler&& h) ->
@@ -757,7 +757,7 @@ struct connection_provider_traits {
  * @ingroup group-connection-concepts
  */
 template <typename T>
-constexpr auto ConnectionProvider = is_connection_provider<std::decay_t<T>>::value;
+inline constexpr auto ConnectionProvider = is_connection_provider<std::decay_t<T>>::value;
 
 #ifdef OZO_DOCUMENTATION
 /**
@@ -851,7 +851,7 @@ struct initiate_async_get_connection {
 };
 } // namespace detail
 
-constexpr get_connection_op<detail::initiate_async_get_connection> get_connection;
+inline constexpr get_connection_op<detail::initiate_async_get_connection> get_connection;
 #endif
 
 

--- a/include/ozo/core/concept.h
+++ b/include/ozo/core/concept.h
@@ -71,7 +71,7 @@ struct has_operator_not<T, std::void_t<decltype(!std::declval<T>())>>
  * @hideinitializer
  */
 template <typename T>
-constexpr auto OperatorNot = has_operator_not<std::decay_t<T>>::value;
+inline constexpr auto OperatorNot = has_operator_not<std::decay_t<T>>::value;
 
 template <typename T, typename Enable = void>
 struct is_output_iterator : std::false_type {};
@@ -93,7 +93,7 @@ struct is_output_iterator<T, typename std::enable_if<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto OutputIterator = is_output_iterator<T>::value;
+inline constexpr auto OutputIterator = is_output_iterator<T>::value;
 
 template <typename T, typename Enable = void>
 struct is_forward_iterator : std::false_type {};
@@ -115,7 +115,7 @@ struct is_forward_iterator<T, typename std::enable_if<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto ForwardIterator = is_forward_iterator<T>::value;
+inline constexpr auto ForwardIterator = is_forward_iterator<T>::value;
 
 template <typename T, typename Enable = void>
 struct is_iterable : std::false_type {};
@@ -135,7 +135,7 @@ struct is_iterable<T, typename std::enable_if<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto Iterable = is_iterable<T>::value;
+inline constexpr auto Iterable = is_iterable<T>::value;
 
 template <typename T, typename Enable = void>
 struct is_insert_iterator : std::false_type {};
@@ -154,7 +154,7 @@ struct is_insert_iterator<T, typename std::enable_if<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto InsertIterator = is_insert_iterator<T>::value;
+inline constexpr auto InsertIterator = is_insert_iterator<T>::value;
 
 /**
  * @brief Boost.Fusion Sequence concept
@@ -164,7 +164,7 @@ constexpr auto InsertIterator = is_insert_iterator<T>::value;
  * @hideinitializer
  */
 template <typename T>
-constexpr auto FusionSequence = boost::fusion::traits::is_sequence<std::decay_t<T>>::value;
+inline constexpr auto FusionSequence = boost::fusion::traits::is_sequence<std::decay_t<T>>::value;
 
 /**
  * @brief Boost.Hana Sequence concept
@@ -174,7 +174,7 @@ constexpr auto FusionSequence = boost::fusion::traits::is_sequence<std::decay_t<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto HanaSequence = boost::hana::Sequence<std::decay_t<T>>::value;
+inline constexpr auto HanaSequence = boost::hana::Sequence<std::decay_t<T>>::value;
 
 /**
  * @brief Boost.Hana Structure concept
@@ -184,7 +184,7 @@ constexpr auto HanaSequence = boost::hana::Sequence<std::decay_t<T>>::value;
  * @hideinitializer
  */
 template <typename T>
-constexpr auto HanaStruct = boost::hana::Struct<std::decay_t<T>>::value;
+inline constexpr auto HanaStruct = boost::hana::Struct<std::decay_t<T>>::value;
 
 /**
  * @brief Boost.Hana String concept
@@ -194,7 +194,7 @@ constexpr auto HanaStruct = boost::hana::Struct<std::decay_t<T>>::value;
  * @hideinitializer
  */
 template <typename T>
-constexpr auto HanaString = decltype(boost::hana::is_a<boost::hana::string_tag>(std::declval<T>()))::value;
+inline constexpr auto HanaString = decltype(boost::hana::is_a<boost::hana::string_tag>(std::declval<T>()))::value;
 
 /**
  * @brief Boost.Hana Tuple concept
@@ -204,7 +204,7 @@ constexpr auto HanaString = decltype(boost::hana::is_a<boost::hana::string_tag>(
  * @hideinitializer
  */
 template <typename T>
-constexpr auto HanaTuple = decltype(boost::hana::is_a<boost::hana::tuple_tag>(std::declval<T>()))::value;
+inline constexpr auto HanaTuple = decltype(boost::hana::is_a<boost::hana::tuple_tag>(std::declval<T>()))::value;
 
 
 template <typename T, typename = std::void_t<>>
@@ -227,7 +227,7 @@ struct is_fusion_adapted_struct<T, std::enable_if_t<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto FusionAdaptedStruct = is_fusion_adapted_struct<std::decay_t<T>>::value;
+inline constexpr auto FusionAdaptedStruct = is_fusion_adapted_struct<std::decay_t<T>>::value;
 
 /**
  * @brief Integral concept
@@ -237,7 +237,7 @@ constexpr auto FusionAdaptedStruct = is_fusion_adapted_struct<std::decay_t<T>>::
  * @hideinitializer
  */
 template <typename T>
-constexpr auto Integral = std::is_integral_v<std::decay_t<T>>;
+inline constexpr auto Integral = std::is_integral_v<std::decay_t<T>>;
 
 /**
  * @brief Floating Point concept
@@ -247,7 +247,7 @@ constexpr auto Integral = std::is_integral_v<std::decay_t<T>>;
  * @hideinitializer
  */
 template <typename T>
-constexpr auto FloatingPoint = std::is_floating_point_v<std::decay_t<T>>;
+inline constexpr auto FloatingPoint = std::is_floating_point_v<std::decay_t<T>>;
 
 template <typename T, std::size_t Size, typename = std::void_t<>>
 struct has_mutable_data : std::false_type {};
@@ -331,7 +331,7 @@ struct is_raw_data_readable : std::bool_constant<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto RawDataWritable = is_raw_data_writable<T>::value;
+inline constexpr auto RawDataWritable = is_raw_data_writable<T>::value;
 
 /**
  * @brief `RawDataReadable` concept
@@ -355,7 +355,7 @@ constexpr auto RawDataWritable = is_raw_data_writable<T>::value;
  * @hideinitializer
  */
 template <typename T>
-constexpr auto RawDataReadable = is_raw_data_readable<T>::value;
+inline constexpr auto RawDataReadable = is_raw_data_readable<T>::value;
 
 template <typename T, typename = std::void_t<>>
 struct is_emplaceable : std::false_type {};
@@ -371,7 +371,7 @@ struct is_emplaceable<T, std::void_t<decltype(std::declval<T&>().emplace())>> : 
  * @hideinitializer
  */
 template <typename T>
-constexpr auto Emplaceable = is_emplaceable<std::decay_t<T>>::value;
+inline constexpr auto Emplaceable = is_emplaceable<std::decay_t<T>>::value;
 
 template <typename T>
 struct is_time_constraint : std::false_type {};
@@ -386,7 +386,7 @@ struct is_time_constraint : std::false_type {};
  * * `ozo::none` --- operation is not restricted in time.
  */
 template <typename T>
-constexpr auto TimeConstraint = is_time_constraint<std::decay_t<T>>::value;
+inline constexpr auto TimeConstraint = is_time_constraint<std::decay_t<T>>::value;
 
 /**
  * @brief Completion token concept
@@ -438,7 +438,7 @@ ozo::connection_type<ConnectionProvider> async_io(
  */
 #ifdef OZO_DOCUMENTATION
 template <typename T>
-constexpr auto CompletionToken = std::false_type;
+inline constexpr auto CompletionToken = std::false_type;
 #endif
 
 /**
@@ -494,7 +494,7 @@ auto Handler = [&] (ozo::error_code ec, auto connection) {
  */
 #ifdef OZO_DOCUMENTATION
 template <typename T>
-constexpr auto Handler = std::false_type;
+inline constexpr auto Handler = std::false_type;
 #endif
 ///@}
 

--- a/include/ozo/core/none.h
+++ b/include/ozo/core/none.h
@@ -25,7 +25,7 @@ template <typename T>
 using is_none = std::is_same<T, none_t>;
 
 template <typename T>
-constexpr auto IsNone = is_none<T>::value;
+inline constexpr auto IsNone = is_none<T>::value;
 
 template <typename T>
 inline constexpr bool operator == (const none_t&, const T&) { return IsNone<T>;}

--- a/include/ozo/core/nullable.h
+++ b/include/ozo/core/nullable.h
@@ -62,7 +62,7 @@ struct allocate_nullable_impl<T*> {
  * @hideinitializer
  */
 template <typename T>
-constexpr auto Nullable = is_nullable<std::decay_t<T>>::value;
+inline constexpr auto Nullable = is_nullable<std::decay_t<T>>::value;
 
 template <typename T>
 struct is_null_impl : std::conditional_t<Nullable<T>,

--- a/include/ozo/core/strong_typedef.h
+++ b/include/ozo/core/strong_typedef.h
@@ -55,7 +55,7 @@ template <typename T, typename Tag>
 struct is_strong_typedef<strong_typedef_wrapper<T, Tag>> : std::true_type {};
 
 template <typename T>
-constexpr auto StrongTypedef = is_strong_typedef<std::decay_t<T>>::value;
+inline constexpr auto StrongTypedef = is_strong_typedef<std::decay_t<T>>::value;
 
 } // namespace ozo
 

--- a/include/ozo/io/binary_query.h
+++ b/include/ozo/io/binary_query.h
@@ -216,7 +216,7 @@ template <typename T>
 using is_binary_query_convertible = typename std::negation<typename std::is_base_of<detail::no_binary_query_conversion, T>::type>::type;
 
 template <typename T>
-constexpr auto is_binary_query_convertible_v = is_binary_query_convertible<T>::value;
+inline constexpr auto is_binary_query_convertible_v = is_binary_query_convertible<T>::value;
 
 /**
  * @brief Concept of a type that is convertible to `ozo::binary_query`
@@ -275,7 +275,7 @@ struct to_binary_query_impl<demo::Query> {
  */
 //! @cond
 template <typename T>
-constexpr auto BinaryQueryConvertible = is_binary_query_convertible_v<std::decay_t<T>>;
+inline constexpr auto BinaryQueryConvertible = is_binary_query_convertible_v<std::decay_t<T>>;
 //! @endcond
 
 template <typename T>

--- a/include/ozo/io/type_traits.h
+++ b/include/ozo/io/type_traits.h
@@ -12,7 +12,7 @@ template <typename T>
 struct has_resize<T, std::void_t<decltype(std::declval<T>().resize(std::size_t()))>> : std::true_type {};
 
 template <typename T>
-constexpr auto Resizable = has_resize<T>::value;
+inline constexpr auto Resizable = has_resize<T>::value;
 
 template <typename T, typename = std::void_t<>>
 struct is_writable : std::false_type {};
@@ -21,7 +21,7 @@ template <typename T>
 struct is_writable<T, std::void_t<decltype(read(std::declval<istream&>(), std::declval<T>()))>> : std::true_type {};
 
 template <typename T>
-constexpr auto Writable = is_writable<T>::value;
+inline constexpr auto Writable = is_writable<T>::value;
 
 template <typename T, typename = std::void_t<>>
 struct is_readable : std::false_type {};
@@ -30,6 +30,6 @@ template <typename T>
 struct is_readable<T, std::void_t<decltype(write(std::declval<ostream&>(), std::declval<T>()))>> : std::true_type {};
 
 template <typename T>
-constexpr auto Readable = is_readable<T>::value;
+inline constexpr auto Readable = is_readable<T>::value;
 
 } // namespace ozo

--- a/include/ozo/query.h
+++ b/include/ozo/query.h
@@ -136,7 +136,7 @@ struct is_query_text<T, std::void_t<
  */
 //! @cond
 template <typename T>
-constexpr auto QueryText = is_query_text<std::decay_t<T>>::value;
+inline constexpr auto QueryText = is_query_text<std::decay_t<T>>::value;
 //! @endcond
 
 template <typename T, typename = hana::when<true>>
@@ -265,7 +265,7 @@ static_assert(ozo::HanaSequence<decltype(ozo::get_query_params(query))>);
  */
 //! @cond
 template <typename T>
-constexpr auto Query = is_query<std::decay_t<T>>::value;
+inline constexpr auto Query = is_query<std::decay_t<T>>::value;
 //! @endcond
 
 /**

--- a/include/ozo/type_traits.h
+++ b/include/ozo/type_traits.h
@@ -92,7 +92,7 @@ using null_oid_t = oid_constant<0>;
  * @brief Constant for empty OID
  * @ingroup group-type_system-constants
  */
-constexpr null_oid_t null_oid;
+inline constexpr null_oid_t null_oid;
 
 /**
  * @brief Indicates if type is PostgreSQL array representation.
@@ -131,7 +131,7 @@ struct is_array : std::false_type {};
  * @hideinitializer
  */
 template <typename T>
-constexpr auto Array = is_array<std::decay_t<T>>::value;
+inline constexpr auto Array = is_array<std::decay_t<T>>::value;
 
 namespace definitions {
 template <typename T>
@@ -205,7 +205,7 @@ using type_traits = decltype(detail::get_type_traits(std::declval<T>()));
  * @hideinitializer
  */
 template <typename T>
-constexpr auto HasDefinition = !std::is_void_v<type_traits<T>>;
+inline constexpr auto HasDefinition = !std::is_void_v<type_traits<T>>;
 
 template <typename T>
 struct is_composite : std::bool_constant<
@@ -225,7 +225,7 @@ struct is_composite : std::bool_constant<
  * @hideinitializer
  */
 template <typename T>
-constexpr auto Composite = is_composite<std::decay_t<T>>::value;
+inline constexpr auto Composite = is_composite<std::decay_t<T>>::value;
 
 /**
  * @brief PostgreSQL size type
@@ -246,7 +246,7 @@ using size_type = std::int32_t;
 template <size_type n>
 struct size_constant : std::integral_constant<size_type, n> {};
 
-constexpr size_constant<-1> null_state_size;
+inline constexpr size_constant<-1> null_state_size;
 
 /**
 * Helpers to make size trait constant
@@ -272,7 +272,7 @@ struct is_built_in<T, std::enable_if_t<
  * @hideinitializer
  */
 template <typename T>
-constexpr bool BuiltIn = is_built_in<std::decay_t<T>>::value;
+inline constexpr bool BuiltIn = is_built_in<std::decay_t<T>>::value;
 
 template <typename T>
 struct is_dynamic_size : std::is_same<typename type_traits<T>::size, dynamic_size> {};
@@ -284,7 +284,7 @@ struct is_dynamic_size : std::is_same<typename type_traits<T>::size, dynamic_siz
  * @hideinitializer
  */
 template <typename T>
-constexpr auto DynamicSize = is_dynamic_size<std::decay_t<T>>::value;
+inline constexpr auto DynamicSize = is_dynamic_size<std::decay_t<T>>::value;
 
 /**
  * @brief Condition indicates if the specified type is has fixed size
@@ -293,7 +293,7 @@ constexpr auto DynamicSize = is_dynamic_size<std::decay_t<T>>::value;
  * @hideinitializer
  */
 template <typename T>
-constexpr auto StaticSize = !DynamicSize<T>;
+inline constexpr auto StaticSize = !DynamicSize<T>;
 
 /**
 * @brief Function returns type name in Postgre SQL.
@@ -500,7 +500,7 @@ bool res = empty(map);
  * @sa oid_map_t, register_types(), set_type_oid(), type_oid(), accepts_oid()
  */
 template <typename T>
-constexpr auto OidMap = is_oid_map<std::decay_t<T>>::value;
+inline constexpr auto OidMap = is_oid_map<std::decay_t<T>>::value;
 
 /**
  * @brief Provides #OidMap implementation for user-defined types.
@@ -538,7 +538,7 @@ constexpr oid_map_t<Ts...> register_types() noexcept {
  */
 using empty_oid_map = std::decay_t<decltype(register_types<>())>;
 
-constexpr empty_oid_map empty_oid_map_c;
+inline constexpr empty_oid_map empty_oid_map_c;
 
 /**
 * Function sets oid for a type in #OidMap.

